File: api/scripts/seed.py
"""
Deterministic dev seed: fixed RNG seed, idempotent via TRUNCATE CASCADE.
Run from api/:  python -m scripts.seed
Requires DATABASE_URL or default postgres.
"""
from __future__ import annotations

import os
import random
import sys
import uuid

sys.path.insert(0, os.path.dirname(os.path.dirname(__file__)))

from sqlalchemy import text

from app.database import SessionLocal
from app.models import (
    Batch,
    Driver,
    InventoryBalance,
    Location,
    Order,
    OrderLine,
    Org,
    Role,
    SKU,
    User,
    UserRole,
    Vehicle,
)
from app.models.enums import RoleName
from app.security import hash_password


SEED = 42
NUM_RETAILERS = 120
NUM_SKUS = 40
NUM_VEHICLES = 100
NUM_DRIVERS = 100


def truncate_all(db) -> None:
    db.execute(
        text(
            """
            TRUNCATE TABLE
              events,
              trip_costs,
              stop_deliveries,
              trip_stops,
              trips,
              allocations,
              order_lines,
              orders,
              inventory_balances,
              batches,
              user_roles,
              users,
              drivers,
              vehicles,
              locations,
              skus,
              roles,
              orgs
            RESTART IDENTITY CASCADE;
            """
        )
    )


def run() -> None:
    random.seed(SEED)
    rng = random.Random(SEED)

    db = SessionLocal()
    try:
        truncate_all(db)
        db.commit()
    except Exception:
        db.rollback()
        raise

    org = Org(name="Getto Demo Org")
    db.add(org)
    db.flush()

    roles: dict[str, Role] = {}
    for rn in RoleName:
        r = Role(name=rn.value)
        db.add(r)
        db.flush()
        roles[rn.value] = r

    def add_user(email: str, name: str, pwd: str, role_names: list[str]) -> User:
        u = User(
            org_id=org.id,
            email=email,
            password_hash=hash_password(pwd),
            full_name=name,
        )
        db.add(u)
        db.flush()
        for rn in role_names:
            db.add(UserRole(user_id=u.id, role_id=roles[rn].id))
        return u

    add_user("admin@getto.demo", "Admin User", "admin123", ["admin"])
    add_user("warehouse@getto.demo", "Warehouse Manager", "warehouse123", ["warehouse_manager"])
    add_user("dispatcher@getto.demo", "Dispatcher", "dispatcher123", ["dispatcher"])
    driver_user = add_user("driver@getto.demo", "Driver One", "driver123", ["driver"])
    add_user("retailer@getto.demo", "Retailer User", "retailer123", ["retailer"])

    factory = Location(
        org_id=org.id,
        type="factory",
        name="Amul Factory",
        address="Mira Bhayandar",
        lat=19.295,
        lng=72.877,
    )
    dairy = Location(
        org_id=org.id,
        type="dairy",
        name="Amul Dairy Staging",
        address="Mira Bhayandar",
        lat=19.30,
        lng=72.88,
    )
    db.add_all([factory, dairy])
    db.flush()

    warehouses: list[Location] = []
    areas = [("Kandivali West", 19.20, 72.85), ("Kandivali East", 19.21, 72.87), ("Bandra", 19.06, 72.83)]
    for i in range(20):
        a = areas[i % 3]
        lat = a[1] + rng.uniform(-0.02, 0.02)
        lng = a[2] + rng.uniform(-0.02, 0.02)
        w = Location(
            org_id=org.id,
            type="warehouse",
            name=f"WH-{i+1:02d} {a[0]}",
            address=f"{a[0]} warehouse",
            lat=lat,
            lng=lng,
        )
        warehouses.append(w)
        db.add(w)
    db.flush()

    retailers: list[Location] = []
    for i in range(NUM_RETAILERS):
        lat = 19.05 + rng.random() * 0.25
        lng = 72.80 + rng.random() * 0.12
        r = Location(
            org_id=org.id,
            type="retailer",
            name=f"Retailer {i+1:04d}",
            address=f"Store {i+1}",
            lat=lat,
            lng=lng,
        )
        retailers.append(r)
        db.add(r)
    db.flush()

    skus: list[SKU] = []
    for i in range(NUM_SKUS):
        s = SKU(
            sku_code=f"SKU-{i+1:04d}",
            name=f"Product {i+1}",
            uom="unit",
            weight_kg=rng.uniform(0.1, 5.0),
            volume_m3=rng.uniform(0.001, 0.05),
        )
        skus.append(s)
        db.add(s)
    db.flush()

    for sku in skus:
        b = Batch(
            sku_id=sku.id,
            batch_code=f"B-{sku.sku_code}",
            qty_produced=rng.uniform(500, 2000),
        )
        db.add(b)
    db.flush()

    for w in warehouses:
        for sku in skus:
            inv = InventoryBalance(
                location_id=w.id,
                sku_id=sku.id,
                on_hand=rng.uniform(200, 2000),
                reserved=0,
            )
            db.add(inv)
    db.flush()

    for i in range(NUM_VEHICLES):
        v = Vehicle(
            org_id=org.id,
            reg_number=f"MH-XX-{i+1:04d}",
            capacity_kg=rng.uniform(800, 3500),
            capacity_m3=rng.uniform(10, 40),
        )
        db.add(v)
    db.flush()

    for i in range(NUM_DRIVERS):
        d = Driver(
            org_id=org.id,
            user_id=driver_user.id if i == 0 else None,
            name=f"Driver {i+1}",
            phone=f"98{rng.randint(10000000, 99999999)}",
        )
        db.add(d)
    db.flush()

    for _ in range(80):
        w = rng.choice(warehouses)
        r = rng.choice(retailers)
        nlines = rng.randint(1, 4)
        o = Order(
            org_id=org.id,
            retailer_location_id=r.id,
            warehouse_location_id=w.id,
            status="pending_allocation",
            priority=rng.randint(0, 3),
        )
        db.add(o)
        db.flush()
        for _ in range(nlines):
            sku = rng.choice(skus)
            db.add(
                OrderLine(
                    order_id=o.id,
                    sku_id=sku.id,
                    qty_requested=rng.uniform(5, 120),
                    qty_allocated=0,
                    status="pending",
                )
            )

    db.commit()
    print("Seed complete: org, roles, users, locations, skus, inventory, vehicles, drivers, orders.")


if __name__ == "__main__":
    run()

File: web/prisma/seed.ts
/**
 * Dev seed — same shape as the former Python seed (smaller optional tweaks via constants).
 * Run: npm run db:seed (from web/)
 */
import { PrismaClient } from "@prisma/client";
import bcrypt from "bcrypt";

const prisma = new PrismaClient();

const SEED = 42;
const NUM_RETAILERS = 120;
const NUM_SKUS = 40;
const NUM_VEHICLES = 100;
const NUM_DRIVERS = 100;

function mulberry32(a: number) {
  return function () {
    let t = (a += 0x6d2b79f5);
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

async function clearAll() {
  await prisma.stopDelivery.deleteMany();
  await prisma.tripCost.deleteMany();
  await prisma.tripStop.deleteMany();
  await prisma.trip.deleteMany();
  await prisma.allocation.deleteMany();
  await prisma.orderLine.deleteMany();
  await prisma.order.deleteMany();
  await prisma.stockTransferLine.deleteMany();
  await prisma.stockTransfer.deleteMany();
  await prisma.inventoryBalance.deleteMany();
  await prisma.batch.deleteMany();
  await prisma.userRole.deleteMany();
  await prisma.driver.deleteMany();
  await prisma.vehicle.deleteMany();
  await prisma.businessEvent.deleteMany();
  await prisma.user.deleteMany();
  await prisma.location.deleteMany();
  await prisma.sKU.deleteMany();
  await prisma.role.deleteMany();
  await prisma.org.deleteMany();
}

async function main() {
  const rng = mulberry32(SEED);

  await clearAll();

  const org = await prisma.org.create({ data: { name: "Smart Login Portal Demo" } });

  const roleNames = [
    "admin",
    "factory_ops",
    "dairy_ops",
    "warehouse_manager",
    "dispatcher",
    "driver",
    "retailer",
    "customer",
  ] as const;
  const roles: Record<string, { id: string }> = {};
  for (const name of roleNames) {
    const r = await prisma.role.create({ data: { name } });
    roles[name] = r;
  }

  async function addUser(
    email: string,
    fullName: string,
    password: string,
    roleKeys: (typeof roleNames)[number][],
  ) {
    const passwordHash = await bcrypt.hash(password, 12);
    const u = await prisma.user.create({
      data: {
        orgId: org.id,
        email,
        passwordHash,
        fullName,
      },
    });
    for (const rn of roleKeys) {
      await prisma.userRole.create({
        data: { userId: u.id, roleId: roles[rn].id },
      });
    }
    return u;
  }

  await addUser("admin@getto.demo", "Admin User", "admin123", ["admin"]);
  await addUser("warehouse@getto.demo", "Warehouse Manager", "warehouse123", ["warehouse_manager"]);
  await addUser("dispatcher@getto.demo", "Dispatcher", "dispatcher123", ["dispatcher"]);
  const driverUser = await addUser("driver@getto.demo", "Driver One", "driver123", ["driver"]);
  await addUser("retailer@getto.demo", "Retailer User", "retailer123", ["retailer"]);

  await prisma.location.create({
    data: {
      orgId: org.id,
      type: "factory",
      name: "Amul Factory",
      address: "Mira Bhayandar",
      lat: 19.295,
      lng: 72.877,
    },
  });
  await prisma.location.create({
    data: {
      orgId: org.id,
      type: "dairy",
      name: "Amul Dairy Staging",
      address: "Mira Bhayandar",
      lat: 19.3,
      lng: 72.88,
    },
  });

  const warehouses: { id: string }[] = [];
  const areas = [
    { name: "Kandivali West", lat: 19.2, lng: 72.85 },
    { name: "Kandivali East", lat: 19.21, lng: 72.87 },
    { name: "Bandra", lat: 19.06, lng: 72.83 },
  ];
  for (let i = 0; i < 20; i++) {
    const a = areas[i % 3];
    const w = await prisma.location.create({
      data: {
        orgId: org.id,
        type: "warehouse",
        name: `WH-${String(i + 1).padStart(2, "0")} ${a.name}`,
        address: `${a.name} warehouse`,
        lat: a.lat + (rng() * 0.04 - 0.02),
        lng: a.lng + (rng() * 0.04 - 0.02),
      },
    });
    warehouses.push(w);
  }

  const retailers: { id: string }[] = [];
  for (let i = 0; i < NUM_RETAILERS; i++) {
    const r = await prisma.location.create({
      data: {
        orgId: org.id,
        type: "retailer",
        name: `Retailer ${String(i + 1).padStart(4, "0")}`,
        address: `Store ${i + 1}`,
        lat: 19.05 + rng() * 0.25,
        lng: 72.8 + rng() * 0.12,
      },
    });
    retailers.push(r);
  }

  const skus: { id: string; skuCode: string }[] = [];
  for (let i = 0; i < NUM_SKUS; i++) {
    const code = `SKU-${String(i + 1).padStart(4, "0")}`;
    const s = await prisma.sKU.create({
      data: {
        skuCode: code,
        name: `Product ${i + 1}`,
        uom: "unit",
        weightKg: 0.1 + rng() * 4.9,
        volumeM3: 0.001 + rng() * 0.049,
      },
    });
    skus.push(s);
  }

  for (const sku of skus) {
    await prisma.batch.create({
      data: {
        skuId: sku.id,
        batchCode: `B-${sku.skuCode}`,
        qtyProduced: 500 + rng() * 1500,
      },
    });
  }

  for (const w of warehouses) {
    for (const sku of skus) {
      await prisma.inventoryBalance.create({
        data: {
          locationId: w.id,
          skuId: sku.id,
          onHand: 200 + rng() * 1800,
          reserved: 0,
        },
      });
    }
  }

  for (let i = 0; i < NUM_VEHICLES; i++) {
    await prisma.vehicle.create({
      data: {
        orgId: org.id,
        regNumber: `MH-XX-${String(i + 1).padStart(4, "0")}`,
        capacityKg: 800 + rng() * 2700,
        capacityM3: 10 + rng() * 30,
      },
    });
  }

  for (let i = 0; i < NUM_DRIVERS; i++) {
    await prisma.driver.create({
      data: {
        orgId: org.id,
        userId: i === 0 ? driverUser.id : undefined,
        name: `Driver ${i + 1}`,
        phone: `98${String(Math.floor(10000000 + rng() * 89999999))}`,
      },
    });
  }

  for (let o = 0; o < 80; o++) {
    const w = warehouses[Math.floor(rng() * warehouses.length)]!;
    const r = retailers[Math.floor(rng() * retailers.length)]!;
    const nlines = 1 + Math.floor(rng() * 4);
    const order = await prisma.order.create({
      data: {
        orgId: org.id,
        retailerLocationId: r.id,
        warehouseLocationId: w.id,
        status: "pending_allocation",
        priority: Math.floor(rng() * 4),
      },
    });
    for (let l = 0; l < nlines; l++) {
      const sku = skus[Math.floor(rng() * skus.length)]!;
      await prisma.orderLine.create({
        data: {
          orderId: order.id,
          skuId: sku.id,
          qtyRequested: 5 + rng() * 115,
          qtyAllocated: 0,
          status: "pending",
        },
      });
    }
  }

  console.log("Seed complete: org, roles, users, locations, skus, inventory, vehicles, drivers, orders.");
}

main()
  .then(() => prisma.$disconnect())
  .catch((e) => {
    console.error(e);
    prisma.$disconnect();
    process.exit(1);
  });
